src/automaton/builder.rs

```diff
@@ -137,89 +181,212 @@ where
+            return;
+        };
+        let transitions_fail = self.matcher.all_transitions(fail_state).collect_vec();
+        for transition in self.matcher.all_constraint_transitions(state).collect_vec() {
+            let target = self.matcher.split_target(transition);
+            self.recently_added.insert(target.0);
+            for &t in &transitions_fail {
+                self.matcher.append_edge(
+                    target,
+                    self.matcher.next_state(t),
+                    self.matcher.constraint(t).cloned(),
+                );
+            }
+        }
     }
 
-    /// Restore the original fail state.
+    fn recently_added_subgraph(
+        &self,
+    ) -> SubgraphRef<'_, StableDiGraph<State, Transition<C>>, NodeIndex> {
+        SubgraphRef::new(&self.matcher.graph, &self.recently_added, true)
+    }
+
+    /// Attempt to merge all recently added nodes with its siblings.
     ///
-    /// The same as `create_fail_state`, but linking to `old_fail_state` instead
-    /// of creating a new state.
-    fn restore_fail_state(&mut self, state: StateID, old_fail_state: StateID) {
-        let last_fail_state = self.find_last_fail_state(state);
-        self.append_edge(last_fail_state, old_fail_state, None);
+    /// This is an optimisation that aims to reduce the number of states
+    /// in the automaton. This leaves the behaviour of the state automaton
+    /// unchanged.
+    fn try_merge_new_nodes(&mut self) {
+        // We try to merge nodes in a (reverse) toposort order, as merging
+        // children first might enable the merge of parent nodes.
+        // We thus start at the newly added nodes that have no newly added descendants
+        let roots = self.recently_added.iter().copied().filter(|&n| {
+            self.matcher
+                .graph
+                .neighbors_directed(n, Direction::Outgoing)
+                .filter(|n| self.recently_added.contains(n))
+                .count()
+                == 0
+        });
+        let mut traverser = OnlineToposort::from_iter(roots);
+        let mut acyclic_checker = AcyclicChecker::with_graph(&self.matcher.graph).unwrap();
+
+        while let Some(node) = traverser.next(self.recently_added_subgraph()) {
+            // Try to merge node with its siblings
+            let state = StateID(node);
+            let Some(first_child) = self
+                .matcher
+                .all_transitions(state)
+                .next()
+                .map(|t| self.matcher.next_state(t))
+            else {
+                continue;
+            };
+            let siblings = self
+                .matcher
+                .incoming_transitions(first_child)
+                .map(|t| self.matcher.parent(t))
+                .unique()
+                .filter(|&n| n != state)
+                // We cannot merge nodes that are reachable from each other
+                .filter(|n| acyclic_checker.path_exists(n.0, state.0, &self.matcher.graph));
```

Comment:
This is not checking "each other" for all pairs.

Is it ok to merge `[n0, n1]` with `state` here?
```mermaid
flowchart
    u --> state
    u --> n0
    state
    n0 ---> n1
    state --> v
    n1 --> v
```